### PR TITLE
MINIFICPP-1140 - Fix VolatileFlowFileRepository

### PR DESCRIPTION
**Pull request by bakaid**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/core/repository/VolatileRepository.h [hunk 1]
```diff
@@ -280,6 +290,19 @@ bool VolatileRepository<T>::Put(T key, const uint8_t *buf, size_t bufLen) {
   logger_->log_debug("VolatileRepository -- put %u %u", current_size_.load(), current_index_.load());
   return true;
 }
+
+template<typename T>
+bool VolatileRepository<T>::MultiPut(const std::vector<std::pair<T, std::unique_ptr<io::DataStream>>>& data) {
+  bool success = true;
+  for (const auto& item : data) {
+    success &= Put(item.first, item.second->getBuffer(), item.second->getSize());
```
Comment (bakaid) [inline libminifi/include/core/repository/VolatileRepository.h]: This does not result in a transacted MultiPut, but I am not sure how that can be achieved with the current architecture of the VolatileRepositories. A rethink of the entire implementation is definitely needed.

File: libminifi/include/core/repository/VolatileFlowFileRepository.h [hunk 2]
```diff
@@ -55,7 +55,9 @@ class VolatileFlowFileRepository : public VolatileRepository<std::string> {
           std::shared_ptr<FlowFileRecord> eventRead = std::make_shared<FlowFileRecord>(shared_from_this(), content_repo_);
           if (eventRead->DeSerialize(reinterpret_cast<const uint8_t *>(purgeItem.data()), purgeItem.size())) {
             std::shared_ptr<minifi::ResourceClaim> newClaim = eventRead->getResourceClaim();
-            content_repo_->remove(newClaim);
+            if (newClaim != nullptr) {
+              content_repo_->removeIfOrphaned(newClaim);
```
Comment (bakaid) [inline libminifi/include/core/repository/VolatileFlowFileRepository.h]: By not choosing the removeIfOrphaned path it
 - could delete a content claim when it still had references to it
 - bypassed the ContentRepository's internal bookkeeping, making it eat more and more RAM.

File: libminifi/include/core/repository/VolatileRepository.h (hunk 1 above)
Comment (bakaid) [reply libminifi/include/core/repository/VolatileRepository.h]: That would still not be in a transaction. An another thread might have updated or deleted and reinserted one of those keys, which we would be deleting with your suggested method.

Comment (arpadboda) [reply libminifi/include/core/repository/VolatileRepository.h]: Okay, that's fair.

File: libminifi/include/core/repository/VolatileRepository.h [hunk 3]
```diff
@@ -280,6 +290,19 @@ bool VolatileRepository<T>::Put(T key, const uint8_t *buf, size_t bufLen) {
...
+    if (!success) {
+      break;
```
[the remark this answers is not in the document]
Comment (bakaid) [reply libminifi/include/core/repository/VolatileRepository.h]: Fair, done.

**Pull request closed without merging**